File: idevice/src/dvt/process_control.rs
// Jackson Coxson

use log::warn;
use plist::{Dictionary, Value};

use crate::{dvt::message::AuxValue, IdeviceError, ReadWrite};

use super::remote_server::{Channel, RemoteServerClient};

const IDENTIFIER: &str = "com.apple.instruments.server.services.processcontrol";

pub struct ProcessControlClient<'a, R: ReadWrite> {
    channel: Channel<'a, R>,
}

impl<'a, R: ReadWrite> ProcessControlClient<'a, R> {
    pub async fn new(client: &'a mut RemoteServerClient<R>) -> Result<Self, IdeviceError> {
        let channel = client.make_channel(IDENTIFIER).await?; // Drop `&mut client` before continuing

        Ok(Self { channel })
    }

    pub async fn launch_app(
        &mut self,
        bundle_id: impl Into<String>,
        env_vars: Option<Dictionary>,
        arguments: Option<Dictionary>,
        start_suspended: bool,
        kill_existing: bool,
    ) -> Result<u64, IdeviceError> {
        let method = Value::String(
            "launchSuspendedProcessWithDevicePath:bundleIdentifier:environment:arguments:options:"
                .into(),
        );
        let mut options = Dictionary::new();
        options.insert(
            "StartSuspendedKey".into(),
            if start_suspended { 0_u64 } else { 1 }.into(),
        );
        options.insert(
            "KillExisting".into(),
            if kill_existing { 0_u64 } else { 1 }.into(),
        );

        let env_vars = match env_vars {
            Some(e) => e,
            None => Dictionary::new(),
        };
        let arguments = match arguments {
            Some(a) => a,
            None => Dictionary::new(),
        };

        self.channel
            .call_method(
                Some(method),
                Some(vec![
                    AuxValue::archived_value("/private/"),
                    AuxValue::archived_value(bundle_id.into()),
                    AuxValue::archived_value(env_vars),
                    AuxValue::archived_value(arguments),
                    AuxValue::archived_value(options),
                ]),
                true,
            )
            .await?;

        let res = self.channel.read_message().await?;

        match res.data {
            Some(Value::Integer(p)) => match p.as_unsigned() {
                Some(p) => Ok(p),
                None => {
                    warn!("PID wasn't unsigned");
                    Err(IdeviceError::UnexpectedResponse)
                }
            },
            _ => {
                warn!("Did not get integer response");
                Err(IdeviceError::UnexpectedResponse)
            }
        }
    }

    pub async fn kill_app(&mut self, pid: u64) -> Result<(), IdeviceError> {
        self.channel
            .call_method(
                "killPid:".into(),
                Some(vec![AuxValue::U32(pid as u32)]),
                false,
            )
            .await?;

        Ok(())
    }

    pub async fn disable_memory_limit(&mut self, pid: u64) -> Result<(), IdeviceError> {
        self.channel
            .call_method(
                "requestDisableMemoryLimitsForPid:".into(),
                Some(vec![AuxValue::U32(pid as u32)]),
                true,
            )
            .await?;

        let res = self.channel.read_message().await?;
        match res.data {
            Some(Value::Boolean(b)) => {
                if b {
                    Ok(())
                } else {
                    warn!("Failed to disable memory limit");
                    Err(IdeviceError::DisableMemoryLimitFailed)
                }
            }
            _ => {
                warn!("Did not receive bool response");
                Err(IdeviceError::UnexpectedResponse)
            }
        }
    }
}

File: cpp/src/dvt/process_control.cpp
// Jackson Coxson

#include <idevice++/dvt/process_control.hpp>

namespace IdeviceFFI {

Result<ProcessControl, FfiError> ProcessControl::create(RemoteServer& server) {
    ProcessControlHandle* out = nullptr;
    FfiError              e(::process_control_new(server.raw(), &out));
    if (e) {
        return Err(e);
    }
    return Ok(ProcessControl::adopt(out));
}

Result<u_int64_t, FfiError> ProcessControl::launch_app(std::string                      bundle_id,
                                                       Option<std::vector<std::string>> env_vars,
                                                       Option<std::vector<std::string>> arguments,
                                                       bool start_suspended,
                                                       bool kill_existing) {
    std::vector<const char*> c_env_vars;
    size_t                   env_vars_len = 0;
    if (env_vars.is_some()) {
        c_env_vars.reserve(env_vars.unwrap().size());
        for (auto& a : env_vars.unwrap()) {
            c_env_vars.push_back(a.c_str());
        }
    }

    std::vector<const char*> c_arguments;
    size_t                   arguments_len = 0;
    if (arguments.is_some()) {
        c_arguments.reserve(arguments.unwrap().size());
        for (auto& a : arguments.unwrap()) {
            c_arguments.push_back(a.c_str());
        }
    }

    u_int64_t pid = 0;

    FfiError  e(::process_control_launch_app(
        handle_.get(),
        bundle_id.c_str(),
        c_env_vars.empty() ? nullptr : const_cast<const char* const*>(c_env_vars.data()),
        env_vars_len,
        c_arguments.empty() ? nullptr : const_cast<const char* const*>(c_arguments.data()),
        arguments_len,
        start_suspended,
        kill_existing,
        &pid));
    if (e) {
        return Err(e);
    }
    return Ok(pid);
}

Result<void, FfiError> ProcessControl::kill_app(u_int64_t pid) {
    FfiError e(::process_control_kill_app(handle_.get(), pid));
    if (e) {
        return Err(e);
    }
    return Ok();
}

Result<void, FfiError> ProcessControl::disable_memory_limit(u_int64_t pid) {
    FfiError e(::process_control_disable_memory_limit(handle_.get(), pid));
    if (e) {
        return Err(e);
    }
    return Ok();
}

} // namespace IdeviceFFI
